src/Features/Core/Portable/InlineMethod/AbstractInlineMethodRefactoringProvider.cs

```diff
@@ -0,0 +1,193 @@
+    {
+        private readonly ISyntaxFacts _syntaxFacts;
+
+        protected abstract Task<SyntaxNode?> GetInvocationExpressionSyntaxNodeAsync(CodeRefactoringContext context);
+
+        /// <summary>
+        /// Check if the <param name="calleeMethodDeclarationSyntaxNode"/> has only one expression or it is using arrow expression.
+        /// </summary>
+        protected abstract bool IsMethodContainsOneStatement(SyntaxNode calleeMethodDeclarationSyntaxNode);
+
+        protected abstract SyntaxNode? GetInlineStatement(SyntaxNode calleeMethodDeclarationSyntaxNode);
+
+        protected AbstractInlineMethodRefactoringProvider(ISyntaxFacts syntaxFacts)
+        {
+            _syntaxFacts = syntaxFacts;
+        }
+
+        public override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            var (document, _, cancellationToken) = context;
+            var calleeMethodInvocationSyntaxNode = await GetInvocationExpressionSyntaxNodeAsync(context).ConfigureAwait(false);
+            if (calleeMethodInvocationSyntaxNode == null)
+            {
+                return;
+            }
+
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            if (semanticModel == null)
+            {
+                return;
+            }
+
+            var methodSymbol = TryGetBestMatchSymbol(semanticModel, calleeMethodInvocationSyntaxNode, cancellationToken);
+            if (methodSymbol == null
+                || methodSymbol.DeclaredAccessibility != Accessibility.Private
+                || !methodSymbol.IsOrdinaryMethod())
+            {
+                return;
+            }
+
+            if (methodSymbol is IMethodSymbol calleeMethodInvocationSymbol)
+            {
+                var calleeMethodDeclarationSyntaxNodes = await Task.WhenAll(calleeMethodInvocationSymbol.DeclaringSyntaxReferences
+                    .Select(reference => reference.GetSyntaxAsync())).ConfigureAwait(false);
+
+                if (calleeMethodDeclarationSyntaxNodes == null || calleeMethodDeclarationSyntaxNodes.Length != 1)
+                {
+                    return;
+                }
+
+                var calleeMethodDeclarationSyntaxNode = calleeMethodDeclarationSyntaxNodes[0];
+
+                if (!IsMethodContainsOneStatement(calleeMethodDeclarationSyntaxNode))
+                {
+                    return;
+                }
+
+                var root = await document.GetSyntaxRootAsync().ConfigureAwait(false);
+                if (root == null)
+                {
+                    return;
+                }
+
+                var codeAction = new CodeAction.DocumentChangeAction(
+                    string.Format(FeaturesResources.Inline_0, calleeMethodInvocationSymbol.ToNameDisplayString()),
+                    cancellationToken => InlineMethodAsync(
+                        document,
+                        semanticModel!,
```

Comment:
do not capture semantic model.

Reply:
I see. So in order to get the semantic model used in code action, a new compilation needs to be generated because the code could be changed right?

Reply:
no.  the code can't actually change.  hwoever, it's important to know that a code action may be computed, but invoked at very different types.  this allows this data to be GC'ed so we're not keepign it alive too long.